File: ai/9/9.py
import sys
'''
8 puzzle using hill climbing
'''
def main():
    data = sys.stdin.read().split()
    n = int(data[0])
    m = int(data[1])
    goal = "123456780"
    start = ""
    idx = 2
    for i in range(n * m):
        start += data[idx]
        idx += 1
    if start == goal:
        print("0")
        return
    def is_solvable(s):
        inv = 0
        for i in range(n * m):
            if s[i] == '0': continue
            for j in range(i + 1, n * m):
                if s[j] == '0': continue
                if s[i] > s[j]:
                    inv += 1
        return inv % 2 == 0
    def heuristic(s):
        h = 0
        for i in range(n * m):
            if s[i] != '0' and int(s[i]) != i + 1:
                h += 1
        return h
    if not is_solvable(start):
        print("No solution")
        return
    di = [-1, 1, 0, 0]
    dj = [0, 0, -1, 1]
    vis = {start}
    parent = {start: ""}
    curr = start
    solved = False
    while True:
        if curr == goal:
            solved = True
            break
        zp = curr.find('0')
        ci, cj = zp // m, zp % m
        currH = heuristic(curr)
        best = curr
        bestH = currH
        for k in range(4):
            ni = ci + di[k]
            nj = cj + dj[k]
            if ni < 0 or ni >= n or nj < 0 or nj >= m:
                continue
            nxt = list(curr)
            nxt[ci * m + cj], nxt[ni * m + nj] = nxt[ni * m + nj], nxt[ci * m + cj]
            nxt = "".join(nxt)
            if nxt in vis:
                continue
            hnext = heuristic(nxt)
            if hnext < bestH:
                bestH = hnext
                best = nxt
        if best == curr:
            break
        vis.add(best)
        parent[best] = curr
        curr = best
    if not solved:
        print("No path...")
    else:
        path = []
        s = goal
        while s:
            path.append(s)
            s = parent[s]
        path.reverse()
        moves = ""
        for i in range(len(path) - 1):
            p1 = path[i].find('0')
            p2 = path[i + 1].find('0')
            i1, j1 = divmod(p1, m)
            i2, j2 = divmod(p2, m)
            if i2 == i1 - 1:
                moves += 'U'
            elif i2 == i1 + 1:
                moves += 'D'
            elif j2 == j1 - 1:
                moves += 'L'
            elif j2 == j1 + 1:
                moves += 'R'
        print("Solution: " + moves)

if __name__ == '__main__':
    main()

'''
OUTPUT:
3 3
1 2 3 
4 0 6 
7 5 8
Solution: DR
'''

File: ai/9/9.cpp
/*
    8 puzzle using Hill climbing
*/
#include <bits/stdc++.h>
using namespace std;

signed main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    int n, m;
    cin >> n >> m;
    const string goal = "123456780";
    string start;
    start.reserve(9);
    for (int i = 0, x; i < n * m; i++) {
        cin >> x;
        start.push_back(char('0' + x));
    }
    if (start == goal) {
        cout << 0 << "\n";
        return 0;
    }
    auto is_solvable = [&](const string &s) {
        int inv = 0;
        for (int i = 0; i < n * m; i++) {
            if (s[i] == '0')
                continue;
            for (int j = i + 1; j < n * m; j++) {
                if (s[j] == '0')
                    continue;
                if (s[i] > s[j])
                    inv++;
            }
        }
        return (inv % 2) == 0;
    };
    auto heuristic = [&](const string &s) {
        int h = 0;
        for (int i = 0; i < n * m; i++) {
            if (s[i] != '0' && s[i] - '0' != i + 1)
                h++;
        }
        return h;
    };
    if (!is_solvable(start)) {
        cout << "No solution\n";
        return 0;
    }
    int di[4] = {-1, 1, 0, 0};
    int dj[4] = {0, 0, -1, 1};
    unordered_set<string> vis;
    unordered_map<string, string> parent;
    vis.insert(start);
    parent[start] = "";
    string curr = start;
    bool solved = false;
    while (true) {
        if (curr == goal) {
            solved = true;
            break;
        }
        int zp = curr.find('0');
        int ci = zp / m, cj = zp % m;
        int currH = heuristic(curr);
        string best = curr;
        int bestH = currH;
        for (int k = 0; k < 4; k++) {
            int ni = ci + di[k], nj = cj + dj[k];
            if (ni < 0 || ni >= n || nj < 0 || nj >= m)
                continue;
            string next = curr;
            swap(next[ci * m + cj], next[ni * m + nj]);
            if (vis.count(next))
                continue; 
            int hnext = heuristic(next);
            if (hnext < bestH) {
                bestH = hnext;
                best = next;
            }
        }
        if (best == curr)
            break;
        vis.insert(best);
        parent[best] = curr;
        curr = best;
    }
    if (!solved) {
        cout << "No path...\n";
    } else {
        vector<string> path;
        for (string s = goal; !s.empty(); s = parent[s])
            path.push_back(s);
        reverse(path.begin(), path.end());
        string moves;
        for (int i = 0; i < (int)path.size() - 1; i++) {
            int p1 = path[i].find('0');
            int p2 = path[i + 1].find('0');
            int i1 = p1 / m, j1 = p1 % m;
            int i2 = p2 / m, j2 = p2 % m;
            if (i2 == i1 - 1)
                moves += 'U';
            else if (i2 == i1 + 1)
                moves += 'D';
            else if (j2 == j1 - 1)
                moves += 'L';
            else if (j2 == j1 + 1)
                moves += 'R';
        }
        cout << "Solution: ";
        cout << moves << "\n";
    }
    return 0;
}